[KEYCLOAK-6650] [KEYCLOAK-6648] Make documentation changes for these JIRAs

iankko opened the pull request:
Create documentation updates for:

- [DONE] [KEYCLOAK-6650]
- [DONE] [KEYCLOAK-6648]

changes. Should be ready for review AFAICT.

Thanks, Jan

Commit:
[KEYCLOAK-6650] Substitute:
* 'redhat-sso72-openshift:1.0' with 'redhat-sso72-openshift:1.1',
* 'ose-v1.4.9' tag with (upcoming) 'ose-v1.4.11' tag

Also update the command to install the updated templates

Signed-off-by: Jan Lieskovsky <jlieskov@redhat.com>

Commit:
[KEYCLOAK-6650] Mention the newly introduced RH-SSO 7.2 x509
application templates on appropriate places

Signed-off-by: Jan Lieskovsky <jlieskov@redhat.com>

Commit:
[KEYCLOAK-6650] Move "Binary Builds" tutorial out of
Getting Started section to Tutorials section

Also rename it to:
"Example Workflow: Create OpenShift Application that Authenticates
Using Red Hat Single Sing-On from Existing Maven Binaries"

Signed-off-by: Jan Lieskovsky <jlieskov@redhat.com>

Commit:
[KEYCLOAK-6650] Rename 'Get Started' section to 'Advanced Concepts'
(we will introduce a new, refactored 'Getting Started' section soon)

Signed-off-by: Jan Lieskovsky <jlieskov@redhat.com>

Commit:
[KEYCLOAK-6650] Bring the refactored 'Getting Started' section
back to the docs

Make it contain the most simplistic example, how to deploy RH-SSO
server

Refactor the 'Advanced Concepts' section to guide:
* How to generate keystores, truststore, and secrets for passthroug
  TLS RH-SSO application templates,
* Also provide example, how the passthrough TLS template can be
  deployed once keystores and secrets are created

Remove the necessary sections from former 'tutorials' content,
that have been used:
* Either in the new 'Getting Started' section, or
* In the new 'Advanced Concepts' section

Signed-off-by: Jan Lieskovsky <jlieskov@redhat.com>

iankko commented:
@matthewhelmke @Crumby @stianst PTAL

(the commits for JIRAs marked WIP to come yet)

matthewhelmke commented on openshift/content/advanced_concepts/advanced_concepts.adoc:
```diff
@@ -0,0 +1,417 @@
+These cover further additional configuration topics, like setup of keystores and truststore for the RH-SSO server, creating an administrator account, overview of available RH-SSO client registration methods, or guidance on how to configure clustering.
+
```
These cover additional configuration topics, such as seting up keystores and a truststore for the RH-SSO server, creating an administrator account, an overview of available RH-SSO client registration methods, and guidance on configuring clustering.

iankko replied:
Thanks, Matthew! Used your suggestion.

matthewhelmke commented on openshift/content/advanced_concepts/advanced_concepts.adoc:
```diff
@@ -0,0 +1,417 @@
+
+[IMPORTANT]
+====
+The RH-SSO application templates using xref:../introduction/introduction.adoc#reencrypt-templates[re-encryption TLS termination] does not *require* / *expect* the aforementioned HTTPS and JGroups keystores and RH-SSO server truststore to be prepared beforehand. They are using OpenShift's internal link:https://docs.openshift.com/container-platform/latest/dev_guide/secrets.html#service-serving-cert ...
```
The RH-SSO application templates using xref:../introduction/introduction.adoc#reencrypt-templates[re-encryption TLS termination] do not *require* or *expect* the aforementioned HTTPS and JGroups keystores and RH-SSO server truststore to be prepared beforehand. The templates use OpenShift's internal link:https://docs.openshift.com/container-platform/latest/dev_guide/secrets.html#service-serving-certificate-secrets[service serving x509 certificate secrets] to automatically create the HTTPS and JGroups keystores. The RH-SSO server truststore is also created automatically, containing the */var/run/secrets/kubernetes.io/serviceaccount/service-ca.crt* CA certificate file, which is used to create these cluster certificates. Moreover, the truststore for the RH-SSO server is pre-populated with the all known, trusted CA certificate files found in the Java system path.

iankko replied:
Again. Updated to the form proposed by you.

matthewhelmke commented on openshift/content/advanced_concepts/advanced_concepts.adoc:
```diff
@@ -0,0 +1,417 @@
+[NOTE]
+====
+For simplicity, in the following command all the provided passwords (the one for the HTTPS keystore, the one for the JGroups keystore, the one for the RH-SSO server truststore, and finally the one for RH-SSO administrator user account) have been chosen to be identical. Although for production environments Red Hat recommends that you use your local password policy to ensure the passwords are diffe ...
+
```
For simplicity, all the provided passwords in the following command (the one for the HTTPS keystore, the one for the JGroups keystore, the one for the RH-SSO server truststore, and finally the one for RH-SSO administrator user account) are identical. For production environments, Red Hat recommends that you use your local password policy to ensure the passwords are different and sufficiently strong. Be aware that *the passwords provided when provisioning the template need to match the passwords provided when creating the keystores*.

iankko replied:
Another time of doing the very same. Thank you for the suggestion!.

matthewhelmke commented on openshift/content/advanced_concepts/advanced_concepts.adoc:
```diff
@@ -0,0 +1,417 @@
+[NOTE]
+====
+Red Hat Single Sign-On allows an initial administrator account creation via the link:https://access.redhat.com/documentation/en-us/red_hat_single_sign-on/7.2/html-single/getting_started_guide/#creating_the_admin_account[Welcome Page] web form. But only if the `Welcome Page` is accessed from a localhost, this method of administrator account creation is not applicable for {xpaasproduct-shortname} i ...
+====
```
Red Hat Single Sign-On allows an initial administrator account to be created via the link:https://access.redhat.com/documentation/en-us/red_hat_single_sign-on/7.2/html-single/getting_started_guide/#creating_the_admin_account[Welcome Page] web form, but only if the `Welcome Page` is accessed from localhost; this method of administrator account creation is not applicable for the {xpaasproduct-shortname} image.

iankko replied:
Replaced with your suggestion.

matthewhelmke commented on openshift/content/advanced_concepts/advanced_concepts.adoc:
```diff
@@ -0,0 +1,417 @@
+====
+The lifespan of the RH-SSO server's administrator account depends upon the the storage type used to store the RH-SSO server's database:
+
+* For an in-memory database mode (*_sso72-https_* and *_sso72-x509-https_* templates) the account exist throughout the lifecycle of the particular RH-SSO pod (stored account data is lost upon pod destruction),
```
...the account exists...

iankko replied:
Replaced two times in the section.

matthewhelmke commented on openshift/content/advanced_concepts/advanced_concepts.adoc:
```diff
@@ -0,0 +1,417 @@
+The lifespan of the RH-SSO server's administrator account depends upon the the storage type used to store the RH-SSO server's database:
+
+* For an in-memory database mode (*_sso72-https_* and *_sso72-x509-https_* templates) the account exist throughout the lifecycle of the particular RH-SSO pod (stored account data is lost upon pod destruction),
+* For an ephemeral database mode (*_sso72-mysql_* and *_sso72-postgresql_* templates) the account exist throughout the lifecycle of the database pod (even if RH-SSO pod is destructed, the stored account data is preserved under the assumption that the database pod is still running),
```
...the account exists...
...even if the RH-SSO pod is...

iankko replied:
Fixed both. Thanks for pointing them out!

matthewhelmke commented on openshift/content/advanced_concepts/advanced_concepts.adoc:
```diff
@@ -0,0 +1,417 @@
+
+* For an in-memory database mode (*_sso72-https_* and *_sso72-x509-https_* templates) the account exist throughout the lifecycle of the particular RH-SSO pod (stored account data is lost upon pod destruction),
+* For an ephemeral database mode (*_sso72-mysql_* and *_sso72-postgresql_* templates) the account exist throughout the lifecycle of the database pod (even if RH-SSO pod is destructed, the stored account data is preserved under the assumption that the database pod is still running),
+* For persistent database mode (*_sso72-mysql-persistent_*, *_sso72-x509-mysql-persistent_*, *_sso72-postgresql-persistent_*, and *_sso72-x509-postgresql-persistent_* templates) the account exists throughout the lifecycle of the persistent medium used to hold the database data. This means that the stored account data is preserved even when both, the RH-SSO and the database pods are destructed.
```
no comma needed after "both"

iankko replied:
Removed.

matthewhelmke commented on openshift/content/advanced_concepts/advanced_concepts.adoc:
```diff
@@ -0,0 +1,417 @@
+[[sso-admin-remote-shell]]
+==== Creating RH-SSO Administrator Account via Remote Shell Session to RH-SSO Pod
+
+Run following commands to create administrator account for the `master` realm of the RH-SSO server, when deploying the {xpaasproduct-shortname} image directly from the image stream (without the xref:../introduction/introduction.adoc#sso-templates[template]), after the RH-SSO application pod has been started:
```
...to create an administrator account... (or "the administrator account" if only one is permitted)

iankko replied:
Used the former form.

matthewhelmke commented on openshift/content/advanced_concepts/advanced_concepts.adoc:
```diff
@@ -0,0 +1,417 @@
+sso-12-pt93n        1/1       Running   0          1m
+sso-mysql-6-d97pf   1/1       Running   0          2m
+----
+. Open a remote shell session to {xpaasproduct-shortname} container.
```
...to the {xpaas...

iankko replied:
add `the`.

matthewhelmke commented on openshift/content/advanced_concepts/advanced_concepts.adoc:
```diff
@@ -0,0 +1,417 @@
++
+[WARNING]
+====
+When restarting the server it is important to restart just the JBoss EAP process within the running RH-SSO container, and not the whole container. Because restarting the whole container recreates it from scratch, without the RH-SSO server administration account for the `master` realm to be created.
```
When restarting the server it is important to restart just the JBoss EAP process within the running RH-SSO container, and not the whole container. This is because restarting the whole container will recreate it from scratch, without the RH-SSO server administration account for the `master` realm.

iankko replied:
Used the form as suggested by you. Thanks for that!

matthewhelmke commented on openshift/content/advanced_concepts/advanced_concepts.adoc:
```diff
@@ -0,0 +1,417 @@
+====
+When restarting the server it is important to restart just the JBoss EAP process within the running RH-SSO container, and not the whole container. Because restarting the whole container recreates it from scratch, without the RH-SSO server administration account for the `master` realm to be created.
+====
+. Log into the `master` realm's administration console of the RH-SSO server using the the credentials created in the steps above. In the browser, navigate to *\http://sso-<project-name>.<hostname>/auth/admin*  for the RH-SSO web server, or to *\https://secure-sso-<project-name>.<hostname>/auth/admin* for the encrypted RH-SSO web server, and specify user name and password used to create the admini ...
```
Log in to the...
...and specify the user name and password...

iankko replied:
Fixed both.

matthewhelmke commented on openshift/content/advanced_concepts/advanced_concepts.adoc:
```diff
@@ -0,0 +1,417 @@
+
+Once deployed, the *_sso72-https_* and *_sso72-x509-https_* templates create a single pod that contains both the database and the RH-SSO servers. The *_sso72-mysql_*, *_sso72-mysql-persistent_*, *_sso72-x509-mysql-persistent_*, *_sso72-postgresql_*, *_sso72-postgresql-persistent_*, and *_sso72-x509-postgresql-persistent_* templates create two pods, one for the database server and one for the RH-S ...
+
+After the RH-SSO web server pod has started, it can be accessed at its custom configured hostnames, or at the default hostnames:
```
...accessed from its... ...or from the default...

iankko replied:
Updated on both places.

matthewhelmke commented on openshift/content/advanced_concepts/advanced_concepts.adoc:
```diff
@@ -0,0 +1,417 @@
+[[SSO-Clients]]
+=== RH-SSO Clients
+
+Clients are RH-SSO entities that request user authentication. A client can be an application requesting RH-SSO to provide user authentication, or it can be making requests for access tokens to start services on behalf of an authenticated user. See the link:https://access.redhat.com/documentation/en-us/red_hat_single_sign-on/7.2/html/server_administration_guide/clients[Managing Clients chapter of  ...
```
...or it can make requests for access...

iankko replied:
Original form replaced.

matthewhelmke commented on openshift/content/introduction/introduction.adoc:
```diff
@@ -2,15 +2,23 @@
 
-For RH-SSO 7.2:
+[[passthrough-templates]]
+* Templates using beforehand prepared HTTPS and JGroups keystores, and truststore for the RH-SSO server. These are securing the TLS communication using link:https://docs.openshift.com/container-platform/latest/architecture/networking/routes.html#passthrough-termination[passthrough TLS termination]:
```
* Templates using HTTPS and JGroups keystores and a truststore for the RH-SSO server, all prepared beforehand. These secure the TLS communication using link:https://docs.openshift.com/container-platform/latest/architecture/networking/routes.html#passthrough-termination[passthrough TLS termination]:

iankko replied:
Fixed. Thank you for the suggestion!

matthewhelmke commented on openshift/content/tutorials/tutorials.adoc:
```diff
@@ -1,174 +1,547 @@
+[[standard-deployments-directory]]
+[NOTE]
+====
+Location of the standard deployments directory depends on the underlying base image, that was used to deploy the application. See the following table:
```
no comma needed after "image"

iankko replied:
Removed.

matthewhelmke commented on openshift/content/tutorials/tutorials.adoc:
```diff
@@ -1,174 +1,547 @@
++
+[NOTE]
+====
+Replace `--image-stream=jboss-eap70-openshift` parameter with the `--image-stream=jboss-eap64-openshift` one in the following oc command to deploy the JSP application on top of JBoss EAP 6.4 for OpenShift image.
```
should "oc" be within backticks? as in ...the following `oc` command...?

iankko replied:
Added them, thanks!

matthewhelmke commented on openshift/content/tutorials/tutorials.adoc:
```diff
@@ -1,174 +1,547 @@
-|_password_
-|===
-. Click *Create* to deploy the application template and start pod deployment. This may take a couple of minutes.
+Access the application in your browser using the URL *\http://eap-app-eap-app-demo.openshift.example.com/app-jsp*. You should see output like on the following image:
```
Access the application in your browser using the URL *\http://eap-app-eap-app-demo.openshift.example.com/app-jsp*. You should see output similar to what you see in the following image:

iankko replied:
Used your form. Thank you.

matthewhelmke commented on openshift/content/tutorials/tutorials.adoc:
```diff
@@ -1,174 +1,547 @@
+You should see the *Message: public* output.
+* Click the *LOGIN* button to be redirected for user authentication to the RH-SSO server instance against the `demo` realm.
++
+Specify username and password of the RH-SSO user configured earlier (`appuser` / `apppassword`). Click *Log in*. The look of the application changes as detailed in the following image:
```
...changes as shown in the following...

iankko replied:
Replaced.

matthewhelmke commented on openshift/content/tutorials/tutorials.adoc:
```diff
@@ -1,174 +1,547 @@
 ====
-This example workflow uses a self-generated CA to provide an end-to-end workflow for demonstration purposes. Accessing the RH-SSO web console will prompt an insecure connection warning. +
-For production environments, Red Hat recommends that you use an SSL certificate purchased from a verified Certificate Authority.
+The `admin` endpoint requires users with `admin` RH-SSO role to invoke properly. Access for the `appuser` is forbidden because they only have `user` role privilege, which allows them to access the `secured` endpoint.
```
To invoke properly, the `admin` endpoint requires users with the `admin` RH-SSO role. Access for the `appuser` is forbidden because they only have the `user` role privilege, which allows them to access the `secured` endpoint.

iankko replied:
Replaced the original with your suggestion. Thank you!

matthewhelmke reviewed (dismissed):
I have some recommended changes to clean up the language a bit.

iankko commented:
@matthewhelmke Thanks for the review, Matthew! (definitely appreciated). Will address them all in one batch.

Commit:
[KEYCLOAK-6650] Address issues pointed out by Matthew during PR
review. Thanks for them, Matthew!

Signed-off-by: Jan Lieskovsky <jlieskov@redhat.com>

Commit:
[KEYCLOAK-6648] Align the definition of HTTPS, JGroups keystores,
and the truststore for the RH-SSO server in the application templates
with their definition in the documentation

Also provide example how to obtain certificate names from keystores

Signed-off-by: Jan Lieskovsky <jlieskov@redhat.com>

iankko commented:
@Crumby PTAL.

Crumby commented on openshift/content/introduction/introduction.adoc:
```diff
@@ -2,15 +2,23 @@
+** *_sso72-postgresql-persistent_*: RH-SSO 7.2 backed by persistent PostgreSQL database on a separate pod.
+
+[[reencrypt-templates]]
+* Templates using OpenShift's internal link:https://docs.openshift.com/container-platform/latest/dev_guide/secrets.html#service-serving-certificate-secrets[service serving x509 certificate secrets] to automatically create the HTTPS keystore used for serving secure content, the JGroups keystore used for securing JGroups communications, and RH-SSO server truststore. These are securing the TLS commu ...
```
Generating JGroups kestores doesn't work yet.

iankko replied:
Works. But uses the deprecated `ENCRYPT` protocol, when `JGROUPS_ENCRYPT_SECRET` variable isn't defined. There's JIRA for ASYM_ENCRYPT support for the future, correct.

iankko replied:
@Crumby Updated the Introduction section to clarify for x509 templates, the JGroups keystore isn't generated. See last commit.

Commit:
[KEYCLOAK-6650] Clarify in the Introduction section, that
for the x509 re-encrypt templates the JGroups keystore isn't
generated, and AUTH protocol is used for cluster traffic
authentication

Signed-off-by: Jan Lieskovsky <jlieskov@redhat.com>

Crumby commented:
LGTM
